File: engine/astra/rpc_handlers/events.py
"""Event/alert broker handlers: transient event ingestion, clustering,
cross-survey association, and public alert-stream polling.

Split out of rpc.py (see that module's docstring for why); nothing here
changed behavior, only location.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from .common import Handler, _workspace_root

from .. import alerts, association, config, events

def _event_root(params: dict[str, Any]) -> Path | None:
    """Use a project workspace for mutable event indexes when supplied."""
    return _workspace_root(params.get("project_id"))


def _handle_event_providers(_params: dict[str, Any]) -> list[dict]:
    return events.providers()


def _handle_event_ingest(params: dict[str, Any]) -> dict:
    if "payload" not in params:
        raise ValueError("events.ingest requires a payload")
    return events.ingest(
        str(params.get("provider", "generic")), params["payload"],
        root=_event_root(params), release=str(params.get("release", "unknown")),
        packet_id=params.get("packet_id"),
        packet_version=str(params.get("packet_version", "1")),
        received_utc=params.get("received_utc"), project_id=params.get("project_id"),
    )


def _handle_event_list(params: dict[str, Any]) -> list[dict]:
    return events.list_events(
        root=_event_root(params), provider=params.get("provider"),
        event_id=params.get("event_id"), limit=int(params.get("limit", 500)),
        packets=bool(params.get("packets", False)),
    )


def _handle_event_get(params: dict[str, Any]) -> dict:
    return events.get_packet(str(params["packet_key"]), root=_event_root(params),
                             include_raw=bool(params.get("include_raw", False)))


def _handle_event_replay(params: dict[str, Any]) -> list[dict]:
    return events.replay(root=_event_root(params), provider=params.get("provider"),
                         event_id=params.get("event_id"),
                         limit=int(params.get("limit", 100)))


def _handle_event_associate(params: dict[str, Any]) -> dict:
    return association.associate_candidates(
        name=str(params.get("name", "default")),
        root=_workspace_root(params.get("project_id")),
        provider=params.get("provider"), event_id=params.get("event_id"),
        radius_arcsec=float(params.get("radius_arcsec", association.DEFAULT_RADIUS_ARCSEC)),
        window_days=float(params.get("window_days", association.DEFAULT_WINDOW_DAYS)),
        allow_unknown_time=bool(params.get("allow_unknown_time", False)),
    )


def _handle_event_graph_correlate(params: dict[str, Any]) -> dict[str, Any]:
    """Pairwise cross-messenger Bayes-factor statistics for ingested events.

    Exposes `association.event_to_event_correlation` -- distinct from
    `events.associate` above, which links an event to a candidate, not one
    event to another. Callers wanting a specific provider/event subset should
    pre-filter with `events.list`/`events.get`; this endpoint always runs
    over `association.fetch_latest_events()`'s full deduplicated view so a
    revised packet cannot silently produce inconsistent results between the
    two association RPCs, matching `fetch_latest_events`'s own contract.
    """
    events_list = association.fetch_latest_events(
        root=_workspace_root(params.get("project_id")), provider=params.get("provider"),
        event_id=params.get("event_id"))
    return {
        "events_checked": len(events_list),
        "pairs": association.event_to_event_correlation(
            events_list,
            window_days=float(params.get("window_days", association.DEFAULT_WINDOW_DAYS)),
            background_window_days=float(params.get("background_window_days", 365.0)),
        ),
    }


def _handle_event_graph_calibrate(params: dict[str, Any]) -> dict[str, Any]:
    """Scrambled-time-slide null calibration for the event-graph Bayes factor."""
    events_list = association.fetch_latest_events(
        root=_workspace_root(params.get("project_id")), provider=params.get("provider"),
        event_id=params.get("event_id"))
    return association.calibrate_event_graph(
        events_list,
        window_days=float(params.get("window_days", association.DEFAULT_WINDOW_DAYS)),
        background_window_days=float(params.get("background_window_days", 365.0)),
        n_trials=int(params.get("n_trials", 200)),
        seed=int(params.get("seed", 42)),
    )


def _handle_alert_providers(_params: dict[str, Any]) -> list[dict]:
    return alerts.providers()


def _handle_alert_status(params: dict[str, Any]) -> dict:
    return alerts.status(_workspace_root(params.get("project_id")) or config.PATHS.root)


def _handle_alert_poll(params: dict[str, Any]) -> dict:
    return alerts.poll(
        str(params["provider"]), endpoint=params.get("endpoint"),
        root=_workspace_root(params.get("project_id")) or config.PATHS.root,
        project_id=params.get("project_id"), cursor=params.get("cursor"),
        limit=int(params.get("limit", 100)), offline=bool(params.get("offline", False)),
        payload=params.get("payload"), params=params.get("params"),
    )


HANDLERS: dict[str, Handler] = {
    "events.providers": _handle_event_providers,
    "events.ingest": _handle_event_ingest,
    "events.list": _handle_event_list,
    "events.get": _handle_event_get,
    "events.replay": _handle_event_replay,
    "events.associate": _handle_event_associate,
    "events.graph.correlate": _handle_event_graph_correlate,
    "events.graph.calibrate": _handle_event_graph_calibrate,
    "alerts.providers": _handle_alert_providers,
    "alerts.status": _handle_alert_status,
    "alerts.poll": _handle_alert_poll,
}

File: src-tauri/src/commands/events.rs
//! Event/alert broker commands: transient event ingestion, clustering,
//! cross-survey association, and public alert-stream polling.
//!
//! Split out of lib.rs (see commands/mod.rs for why); nothing here changed
//! behavior, only location.

use std::sync::Arc;

use serde_json::{json, Value};

use crate::engine::Engine;

use super::call_blocking;

#[tauri::command]
pub(crate) async fn engine_event_providers(state: tauri::State<'_, Arc<Engine>>) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.providers", json!({})).await
}


#[tauri::command]
pub(crate) async fn engine_event_ingest(
    state: tauri::State<'_, Arc<Engine>>,
    provider: String,
    payload: Value,
    release: Option<String>,
    packet_id: Option<String>,
    packet_version: Option<String>,
    received_utc: Option<String>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.ingest", json!({
        "provider": provider,
        "payload": payload,
        "release": release,
        "packet_id": packet_id,
        "packet_version": packet_version,
        "received_utc": received_utc,
        "project_id": project_id,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_events(
    state: tauri::State<'_, Arc<Engine>>,
    provider: Option<String>,
    event_id: Option<String>,
    limit: Option<u32>,
    packets: Option<bool>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.list", json!({
        "provider": provider,
        "event_id": event_id,
        "limit": limit.unwrap_or(500),
        "packets": packets.unwrap_or(false),
        "project_id": project_id,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_event_packet(
    state: tauri::State<'_, Arc<Engine>>,
    packet_key: String,
    include_raw: Option<bool>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.get", json!({
        "packet_key": packet_key,
        "include_raw": include_raw.unwrap_or(false),
        "project_id": project_id,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_event_replay(
    state: tauri::State<'_, Arc<Engine>>,
    provider: Option<String>,
    event_id: Option<String>,
    limit: Option<u32>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.replay", json!({
        "provider": provider,
        "event_id": event_id,
        "limit": limit.unwrap_or(100),
        "project_id": project_id,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_event_associate(
    state: tauri::State<'_, Arc<Engine>>,
    name: Option<String>,
    provider: Option<String>,
    event_id: Option<String>,
    radius_arcsec: Option<f64>,
    window_days: Option<f64>,
    allow_unknown_time: Option<bool>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "events.associate", json!({
        "name": name.unwrap_or_else(|| "default".into()),
        "provider": provider,
        "event_id": event_id,
        "radius_arcsec": radius_arcsec.unwrap_or(30.0),
        "window_days": window_days.unwrap_or(30.0),
        "allow_unknown_time": allow_unknown_time.unwrap_or(false),
        "project_id": project_id,
    })).await
}


#[tauri::command]
pub(crate) async fn engine_alert_providers(state: tauri::State<'_, Arc<Engine>>) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "alerts.providers", json!({})).await
}


#[tauri::command]
pub(crate) async fn engine_alert_status(
    state: tauri::State<'_, Arc<Engine>>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "alerts.status", json!({ "project_id": project_id })).await
}


#[tauri::command]
pub(crate) async fn engine_alert_poll(
    state: tauri::State<'_, Arc<Engine>>,
    provider: String,
    endpoint: Option<String>,
    cursor: Option<String>,
    limit: Option<u32>,
    offline: Option<bool>,
    payload: Option<Value>,
    params: Option<Value>,
    project_id: Option<String>,
) -> Result<Value, String> {
    call_blocking(Arc::clone(&state), "alerts.poll", json!({
        "provider": provider,
        "endpoint": endpoint,
        "cursor": cursor,
        "limit": limit.unwrap_or(100),
        "offline": offline.unwrap_or(false),
        "payload": payload,
        "params": params,
        "project_id": project_id,
    })).await
}


